README.md
Description
===========

This is a modified version of opscode chef-client v2.2.2. 
See changelog for my changes.


Requirements
============

Chef 0.10.10 and Ohai 0.6.12 are required due to the use of
`platform_family`.

Platforms
---------

The following platforms are supported by this cookbook, meaning that the recipes run on these platforms without error.

* Debian family (Debian, Ubuntu etc)
* Red Hat family (Redhat, CentOS, Oracle etc)
* Fedora family
* SUSE distributions (OpenSUSE, SLES, etc)
* ArchLinux
* FreeBSD
* Mac OS X
* Mac OS X Server

Opscode Cookbooks
-----------------

The `chef-client::cron` recipe requires the `cron_d` LWRP from Opscode's cron cookbook, so that is a dependency, otherwise the cron recipe will fail to converge.

Other cookbooks can be used with this cookbook but they are not explicitly required. The default settings in this cookbook do not require their use. The other cookbooks (on community.opsocde.com) are:

* bluepill
* daemontools
* runit

See __USAGE__ below.

Attributes
==========

* `node["chef_client"]["interval"]` - Sets `Chef::Config[:interval]` via command-line option for number of seconds between chef-client daemon runs. Default 1800.
* `node["chef_client"]["splay"]` - Sets `Chef::Config[:splay]` via command-line option for a random amount of seconds to add to interval. Default 20.
* `node["chef_client"]["log_dir"]` - Sets directory used in `Chef::Config[:log_location]` via command-line option to a location where chef-client should log output. Default "/var/log/chef".
* `node["chef_client"]["conf_dir"]` - Sets directory used via command-line option to a location where chef-client search for the client config file . Default "/etc/chef".
* `node["chef_client"]["bin"]` - Sets the full path to the `chef-client` binary.  Mainly used to set a specific path if multiple versions of chef-client exist on a system or the bin has been installed in a non-sane path. Default "/usr/bin/chef-client".
* `node["chef_client"]["server_url"]` - Sets `Chef::Config[:chef_server_url]` in the config file to the Chef Server URI. Default "http://localhost:4000". See __USAGE__.
* `node["chef_client"]["validation_client_name"]` - Sets `Chef::Config[:validation_client_name]` in the config file to the name of the validation client. Default "chef-validator". See __USAGE__.
* `node["chef_client"]["init_style"]` - Sets up the client service based on the style of init system to use. Default is based on platform and falls back to "none". See __USAGE__.
* `node["chef_client"]["run_path"]` - Directory location where chef-client should write the PID file. Default based on platform, falls back to "/var/run".
* `node["chef_client"]["cache_path"]` - Directory location for `Chef::Config[:file_cache_path]` where chef-client will cache various files. Default is based on platform, falls back to "/var/chef/cache".
* `node["chef_client"]["backup_path"]` - Directory location for `Chef::Config[:file_backup_path]` where chef-client will backup templates and cookbook files. Default is based on platform, falls back to "/var/chef/backup".
* `node["chef_client"]["cron"]["minute"]` - The hour that chef-client will run as a cron task, only applicable if the you set "cron" as the "init_style"
* `node["chef_client"]["cron"]["hour"]` - The hour that chef-client will run as a cron task, only applicable if the you set "cron" as the "init_style"
* `node["chef_client"]["cron"]["environment_variables"]` - Environment variables to pass to chef-client's execution (e.g. SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt chef-client)
* `node["chef_client"]["cron"]["log_file"]` - Location to capture the chef-client output.
* `node["chef_client"]["cron"]["use_cron_d"]` - If true, use the "cron_d" LWRP (https://github.com/opscode-cookbooks/cron). If false (default), use the cron resource built-in to Chef.
* `node["chef_client"]["load_gems"]` - Hash of gems to load into chef via the client.rb file
* `node["chef_client"]["report_handlers"]` - Array of Hashes that contain a class and arguments element appended in the client.rb file. ex: `{:class => "MyHandler", :arguments => [true]}
* `node["chef_client"]["exception_handlers"]` - Array of Hashes that contain a class and arguments element appended in the client.rb file. ex: `{:class => "MyHandler", :arguments => [true]}
* `node["chef_client"]["checksum_cache_skip_expires"]` - true or false value used in the `cache_options` section of the client.rb file.
* `node["chef_client"]["checksum_cache_path"]` -  file system path used in the `cache_options` section of the client.rb file.
* `node["chef_client"]["launchd_mode"]` - (Only for Mac OS X) if set to "daemon", runs chef-client with `-d` and `-s` options; defaults to "interval"
* `node["chef_client"]["daemon_options"]` - An array of additional options to pass to the chef-client service, empty by default, and must be an array if specified.
* `node["ohai"]["disabled_plugins"]` - An array of ohai plugins to disable, empty by default, and must be an array if specified.

Recipes
=======

This section describes the recipes in the cookbook and how to use them in your environment.

config
------

Sets up the `/etc/chef/client.rb` config file from a template and reloads the configuration for the current chef-client run.

service
-------

Use this recipe on systems that should have a `chef-client` daemon running, such as when Knife bootstrap was used to install Chef on a new system.

This recipe sets up the `chef-client` service depending on the `init_style` attribute (see above). The following init styles are supported:

* init - uses the init script included in the chef gem, supported on debian and redhat family distributions.
* upstart - uses the upstart job included in the chef gem, supported on ubuntu.
* arch - uses the init script included in this cookbook for ArchLinux, supported on arch.
* runit - sets up the service under runit, supported on ubuntu, debian and gentoo.
* bluepill - sets up the service under bluepill. As bluepill is a pure ruby process monitor, this should work on any platform.
* daemontools - sets up the service under daemontools, supported on debian, ubuntu and arch
* launchd - sets up the service under launchd, supported on Mac OS X & Mac OS X Server. (this requires Chef >= 0.10.10)
* bsd - prints a message about how to update BSD systems to enable the chef-client service, supported on Free/OpenBSD.

default
-------

Includes the `chef-client::service` recipe by default.

delete_validation
-----------------

Use this recipe to delete the validation certificate (default `/etc/chef/validation.pem`) when using a `chef-client` after the client has been validated and authorized to connect to the server.

Beware if using this on your Chef Server. First copy the validation.pem certificate file to another location, such as your knife configuration directory (`~/.chef`) or [Chef Repository](http://wiki.opscode.com/display/chef/Chef+Repository).

cron
----

Use this recipe to run chef-client as a cron job rather than as a
service. The cron job runs after random delay that is between 0 and 90
seconds to ensure that the chef-clients don't attempt to connect to
the chef-server at the exact same time. You should set
node["chef_client"]["init_style"] = "none" when you use this mode but
it is not required.



USAGE
=====

Create a `base` role that will represent the base configuration for any system that includes managing aspects of the chef-client. Add recipes to the run list of the role, customize the attributes, and apply the role to nodes. For example, the following role (Ruby DSL) will set the init style to `init`, delete the validation certificate (as the client would already be authenticated) and set up the chef-client as a service using the init style.

    name "base"
    description "Base role applied to all nodes"
    override_attributes(
      "chef_client" => {
        "init_style" => "init"
      }
    )
    run_list(
      "recipe[chef-client::delete_validation]",
      "recipe[chef-client::config]",
      "recipe[chef-client::service]"
    )

The `chef-client::config` recipe is only required with init style `init` (default setting for the attribute on debian/redhat family platforms, because the init script doesn't include the `pid_file` option which is set in the config.

The default Chef Server will be `http://localhost:4000` which is the `Chef::Config[:chef_server_url]` default value. To use the config recipe with the Opscode Platform, for example, add the following to the `override_attributes`

    override_attributes(
      "chef_client" => {
        "server_url" => "https://api.opscode.com/organizations/ORGNAME",
        "validation_client_name" => "ORGNAME-validator"
      }
    )

Where ORGNAME is your Opscode Platform organization name. Be sure to add these attributes to the role if modifying per the section below.

You can also set all of the `Chef::Config` http proxy related settings.  By default Chef will not use a proxy.

    override_attributes(
      "chef_client" => {
        "http_proxy" => "http://proxy.vmware.com:3128",
        "https_proxy" => "http://proxy.vmware.com:3128",
        "http_proxy_user" => "my_username",
        "http_proxy_pass" => "Awe_some_Pass_Word!",
        "no_proxy" => "*.vmware.com,10.*"
      }
    )

Alternate Init Styles
---------------------

The alternate init styles available are:

* runit
* bluepill
* daemontools
* none -- should be specified if you are running chef-client as cron job

For usage, see below.

# Runit

To use runit, download the cookbook from the cookbook site.

    knife cookbook site vendor runit -d

Change the `init_style` to runit in the base role and add the runit recipe to the role's run list:

    name "base"
    description "Base role applied to all nodes"
    override_attributes(
      "chef_client" => {
        "init_style" => "runit"
      }
    )
    run_list(
      "recipe[chef-client::delete_validation]",
      "recipe[runit]",
      "recipe[chef-client]"
    )

The `chef-client` recipe will create the chef-client service configured with runit. The runit run script will be located in `/etc/sv/chef-client/run`. The output log will be in the runit service directory, `/etc/sv/chef-client/log/main/current`.

# Bluepill

To use bluepill, download the cookbook from the cookbook site.

    knife cookbook site vendor bluepill -d

Change the `init_style` to runit in the base role and add the bluepill recipe to the role's run list:

    name "base"
    description "Base role applied to all nodes"
    override_attributes(
      "chef_client" => {
        "init_style" => "bluepill"
      }
    )
    run_list(
      "recipe[chef-client::delete_validation]",
      "recipe[bluepill]",
      "recipe[chef-client]"
    )

The `chef-client` recipe will create the chef-client service configured with bluepill. The bluepill "pill" will be located in `/etc/bluepill/chef-client.pill`. The output log will be to client.log file in the `node["chef_client"]["log_dir"]` location, `/var/log/chef/client` by default.

# Daemontools

To use daemontools, download the cookbook from the cookbook site.

    knife cookbook site vendor daemontools -d

Change the `init_style` to runit in the base role and add the daemontools recipe to the role's run list:

    name "base"
    description "Base role applied to all nodes"
    override_attributes(
      "chef_client" => {
        "init_style" => "daemontools"
      }
    )
    run_list(
      "recipe[chef-client::delete_validation]",
      "recipe[daemontools]",
      "recipe[chef-client]"
    )

The `chef-client` recipe will create the chef-client service configured under daemontools. It uses the same sv run scripts as the runit recipe. The run script will be located in `/etc/sv/chef-client/run`. The output log will be in the daemontools service directory, `/etc/sv/chef-client/log/main/current`.

# Launchd

On Mac OS X and Mac OS X Server, the default service implementation is "launchd". Launchd support for the service resource is only supported from Chef 0.10.10 onwards.
An error message will be logged if you try to use the launchd service for chef-client on a Chef version that does not contain this launchd support.

Since launchd can run a service in interval mode, by default chef-client is not started in daemon mode like on Debian or Ubuntu. Keep this in mind when you look at your process list and check for a running chef process! If you wish to run chef-client in daemon mode, set attribute `chef_client.launchd_mode` to "daemon".

Templates
=========

chef-client.pill.erb
--------------------

Bluepill configuration for the chef-client service.

client.rb.erb
-------------

Configuration for the client, lands in directory specified by `node["chef_client"]["conf_dir"]` (`/etc/chef/client.rb` by default).

`sv-chef-client-*run.erb`
-------------------------

Runit and Daemontools run script for chef-client service and logs.

Logs will be located in the `node["chef_client"]["log_dir"]`.

com.opscode.chef-client.plist
-----------------------------

Launchd configuration file for chef-client as a true Mac service. The template accepts the `node["chef_client"]["interval"]` value.

License and Author
==================

Author:: Joshua Timberman (<joshua@opscode.com>)
Author:: Seth Chisamore (<schisamo@opscode.com>)
Copyright:: 2010-2012, Opscode, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and


CHANGELOG.md
## v2.2.3:
* This is not an official GetChef (Opscode) v2.2.3
* Updated the chef-client init for rhel to actually kill all chef process, the original init script fails to stop the chef-client process if it is hanging
* Updated /var/log/chef to be 775, in order for monitoring accounts such as icinga to monitor the log file and it's contents  
* Updated the chef-client action to :start and :enable instead of :restart

## v2.2.2:

* [COOK-2393] - chef-client::delete_validation checks for chef-server
  in the path, on chef 11, needs to check for chef-server-ctl
* [COOK-2410] - chef-client::service doesn't always start the
  chef-client daemon
* [COOK-2413] - Deprecation warning when using Chef::Mixin::Language
  in chef-client cookbook under chef 11.x
* [COOK-2446] - Typo: the chef-client executable has a hyphen
* [COOK-2492] - Ruby System("") call that includes an '&' on Ubuntu
  has odd behavior.
* [COOK-2536] - On Freebsd - chef-client group values in helper
  library should be set to "wheel" vs [ "wheel" ]

## v2.2.0:

* [COOK-2317] - Provide the ability to add disabled ohai plugins in a
  managed chef config
* [COOK-2255] - Chef-Client Cookbook init.d script under ubuntu

## v2.1.10:

* [COOK-2316] - Permissions for SMF init type break Solaris 10

## v2.1.8:

* [COOK-2192] - Add option to use cron_d resource for cron management
* [COOK-2261] - pin runit dependency

## v2.1.6:

* [COOK-1978] - make cron output location configurable
* [COOK-2169] - use helper library to make path permissions consistent
* [COOK-2170] - test filename cleanup (dev repository only)

## v2.1.4:

* [COOK-2108] - corrected Chef and Ohai version requirements in README

## v2.1.2:

* [COOK-2071] - chef-client breaks on value_for_platform_family b/c of
  unneeded version
* [COOK-2072] - chef-client on mac should not attempt to create
  directory nil
* [COOK-2086] - Allow the passing of an enviornment variables to
  node['chef-client']['bin']
* [COOK-2092] - chef-client run fails because quotes in log_path cause
  File resource to fail

## v2.1.0:

* [COOK-1755] - Don't delete the validation key on systems that have a
  'chef-server' binary in the default $PATH
* [COOK-1898] - Support Handlers and Cache Options with Attributes
* [COOK-1923] - support chef-client::cron on Solaris/SmartOS
* [COOK-1924] - use splay for size of random offset in
  chef-client::cron
* [COOK-1927] - unknown node[:fqdn] prevents bootstrap if
  chef-client::cron is in runlist
* [COOK-1951] - Add an attribute for additional daemon options to pass
  to the chef-client service
* [COOK-2004] - in attributes, "init" style claims to handle fedora,
  but service.rb missing a clause
* [COOK-2017] - Support alternate chef-client locations in Mac OS X
  Launchd service plist
* [COOK-2052] - Log files are set to insecure default

## v2.0.2:

* Remove a stray comma that caused syntax error on some versions of Ruby.

## v2.0.0:

This version uses platform_family attribute, making the cookbook incompatible
with older versions of Chef/Ohai, hence the major version bump.

* [COOK-635] - Allow configuration of path to chef-client binary in init script
* [COOK-985] - set correct permissions on run and log directory for chef-servers using this cookbook
* [COOK-1379] - Register chef-client as a launchd service on Mac OS X (Server)
* [COOK-1574] - config recipe doesn't work on Windows
* [COOK-1586] - add SmartOS support
* [COOK-1633] - chef-client doesn't recognise Oracle Linux, a Redhat family member
* [COOK-1634] - chef-client init is missing for Scientific Linux
* [COOK-1664] - corrected permissions in cron recipe (related to COOK-985)
* [COOK-1729] - support windows task
* [COOK-1788] - `init_style` upstart only works on Ubuntu
* [COOK-1861] - Minor styling fix for consistency in chef-client
* [COOK-1862] - add `name` attribute to metadata.rb

## v1.2.0:

This version of the cookbook also adds minitest and test-kitchen
support.

* [COOK-599] - chef-client::config recipe breaks folder permissions of
  chef-server::rubygems-install recipe on same node
* [COOK-1111] - doesn't work out of the box with knife bootstrap
  windows
* [COOK-1161] - allow setting log file and environment in client.rb
* [COOK-1203] - allow PATH setting for cron
* [COOK-1254] - service silently fails on ubuntu 12.04 with ruby 1.9.3
* [COOK-1309] - cron recipe expects SANE_PATHS constant
* [COOK-1345] - preformat log location before sending to template
* [COOK-1377] - allow client.rb to require gems
* [COOK-1419] - add init script for SUSE
* [COOK-1463] - Add verbose_logging knob for config recipe, client.rb
  template
* [COOK-1506] - set an attribute for server_url
* [COOK-1566] - remove random splay for unique sleep number

## v1.1.4:

* [COOK-599] - don't break folder permissions if chef-server recipe is present

## v1.1.2:

* [COOK-1039] - support mac_os_x_server

## v1.1.0:

* [COOK-909] - trigger upstart on correct event
* [COOK-795] - add windows support with winsw
* [COOK-798] - added recipe to run chef-client as a cron job
* [COOK-986] - don't delete the validation.pem if chef-server recipe
  is detected

## v1.0.4:

* [COOK-670] - Added Solaris service-installation support for chef-client cookbook.
* [COOK-781] - chef-client service recipe fails with chef 0.9.x

## v1.0.2:

* [CHEF-2491] init scripts should implement reload

## v1.0.0:

* [COOK-204] chef::client pid template doesn't match package expectations
* [COOK-491] service config/defaults should not be pulled from Chef gem
* [COOK-525] Tell bluepill to daemonize chef-client command
* [COOK-554] Typo in backup_path
* [COOK-609] chef-client cookbook fails if init_type is set to upstart and chef is installed from deb
* [COOK-635] Allow configuration of path to chef-client binary in init script


metadata.rb
name              "chef-client"
maintainer        "Reigner S. Yrastorza"
maintainer_email  "reigner@yrastorza.us"
license           "Apache 2.0"
description       "Forked version of GetChef chef-client cookbook"
long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version           "2.2.3"
recipe            "chef-client", "Includes the service recipe by default."
recipe            "chef-client::config", "Configures the client.rb from a template."
recipe            "chef-client::service", "Sets up a client daemon to run periodically"
recipe            "chef-client::delete_validation", "Deletes validation.pem after client registers"
recipe            "chef-client::cron", "Runs chef-client as a cron job rather than as a service"

%w{ ubuntu debian redhat centos fedora oracle suse freebsd openbsd mac_os_x mac_os_x_server windows }.each do |os|
  supports os
end

suggests "bluepill"
suggests "daemontools"
suggests "runit", "<= 0.16.2"
depends "cron"


recipes/service.rb
#
# Author:: Joshua Timberman (<joshua@opscode.com>)
# Author:: Seth Chisamore (<schisamo@opscode.com>)
# Cookbook Name:: chef-client
# Recipe:: service
#
# Copyright 2009-2011, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

class ::Chef::Recipe
  include ::Opscode::ChefClient::Helpers
end

require 'chef/version_constraint'
require 'chef/exceptions'

root_group = value_for_platform_family(
  ["openbsd", "freebsd", "mac_os_x"] => "wheel",
  "default" => "root"
)

if node["platform"] == "windows"
    existence_check = :exists?
# Where will also return files that have extensions matching PATHEXT (e.g.
# *.bat). We don't want the batch file wrapper, but the actual script.
    which = 'set PATHEXT=.exe & where'
    Chef::Log.debug "Using exists? and 'where', since we're on Windows"
else
    existence_check = :executable?
    which = 'which'
    Chef::Log.debug "Using executable? and 'which' since we're on Linux"
end

# COOK-635 account for alternate gem paths
# try to use the bin provided by the node attribute
if ::File.send(existence_check, node["chef_client"]["bin"])
  client_bin = node["chef_client"]["bin"]
  Chef::Log.debug "Using chef-client bin from node attributes: #{client_bin}"
# search for the bin in some sane paths
elsif Chef::Client.const_defined?('SANE_PATHS') && (chef_in_sane_path=Chef::Client::SANE_PATHS.map{|p| p="#{p}/chef-client";p if ::File.send(existence_check, p)}.compact.first) && chef_in_sane_path
  client_bin = chef_in_sane_path
  Chef::Log.debug "Using chef-client bin from sane path: #{client_bin}"
# last ditch search for a bin in PATH
elsif (chef_in_path=%x{#{which} chef-client}.chomp) && ::File.send(existence_check, chef_in_path)
  client_bin = chef_in_path
  Chef::Log.debug "Using chef-client bin from system path: #{client_bin}"
else
  raise "Could not locate the chef-client bin in any known path. Please set the proper path by overriding node['chef_client']['bin'] in a role."
end

node.set["chef_client"]["bin"] = client_bin

# libraries/helpers.rb method to DRY directory creation resources
create_directories

case node["chef_client"]["init_style"]
when "init"

  #argh?
  dist_dir, conf_dir = value_for_platform_family(
    ["debian"] => ["debian", "default"],
    ["fedora"] => ["redhat", "sysconfig"],
    ["rhel"] => ["redhat", "sysconfig"],
    ["suse"] => ["suse", "sysconfig"]
  )

  template "/etc/init.d/chef-client" do
    source "#{dist_dir}/init.d/chef-client.erb"
    mode 0755
    variables(
      :client_bin => client_bin
    )
    notifies :start, "service[chef-client]", :delayed
  end

  template "/etc/#{conf_dir}/chef-client" do
    source "#{dist_dir}/#{conf_dir}/chef-client.erb"
    mode 0644
    notifies :start, "service[chef-client]", :delayed
  end

  service "chef-client" do
    supports :status => true, :restart => true
    action [:enable, :start]
  end

when "smf"
  directory node['chef_client']['method_dir'] do
    action :create
    owner "root"
    group "bin"
    mode "0755"
    recursive true
  end

  local_path = ::File.join(Chef::Config[:file_cache_path], "/")
  template "#{node['chef_client']['method_dir']}/chef-client" do
    source "solaris/chef-client.erb"
    owner "root"
    group "root"
    mode "0755"
    notifies :restart, "service[chef-client]"
  end

  template(local_path + "chef-client.xml") do
    source "solaris/manifest.xml.erb"
    owner "root"
    group "root"
    mode "0644"
    notifies :run, "execute[load chef-client manifest]", :immediately
  end

  execute "load chef-client manifest" do
    action :nothing
    command "svccfg import #{local_path}chef-client.xml"
    notifies :restart, "service[chef-client]"
  end

  service "chef-client" do
    action [:enable, :start]
    provider Chef::Provider::Service::Solaris
  end

when "upstart"

  upstart_job_dir = "/etc/init"
  upstart_job_suffix = ".conf"

  case node["platform"]
  when "ubuntu"
    if (8.04..9.04).include?(node["platform_version"].to_f)
      upstart_job_dir = "/etc/event.d"
      upstart_job_suffix = ""
    end
  end

  template "#{upstart_job_dir}/chef-client#{upstart_job_suffix}" do
    source "debian/init/chef-client.conf.erb"
    mode 0644
    variables(
      :client_bin => client_bin
    )
    notifies :restart, "service[chef-client]", :delayed
  end

  service "chef-client" do
    provider Chef::Provider::Service::Upstart
    action [:enable,:start]
  end

when "arch"

  template "/etc/rc.d/chef-client" do
    source "rc.d/chef-client.erb"
    mode 0755
    variables(
      :client_bin => client_bin
    )
    notifies :restart, "service[chef-client]", :delayed
  end

  template "/etc/conf.d/chef-client.conf" do
    source "conf.d/chef-client.conf.erb"
    mode 0644
    notifies :restart, "service[chef-client]", :delayed
  end

  service "chef-client" do
    action [:enable, :start]
  end

when "runit"

  include_recipe "runit"
  runit_service "chef-client"

when "bluepill"

  directory node["chef_client"]["run_path"] do
    recursive true
    owner "root"
    group root_group
    mode 0755
  end

  include_recipe "bluepill"

  template "#{node["bluepill"]["conf_dir"]}/chef-client.pill" do
    source "chef-client.pill.erb"
    mode 0644
    notifies :restart, "bluepill_service[chef-client]", :delayed
  end

  bluepill_service "chef-client" do
    action [:enable,:load,:start]
  end

when "daemontools"

  include_recipe "daemontools"

  directory "/etc/sv/chef-client" do
    recursive true
    owner "root"
    group root_group
    mode 0755
  end

  daemontools_service "chef-client" do
    directory "/etc/sv/chef-client"
    template "chef-client"
    action [:enable,:start]
    log true
  end

when "winsw"

  directory node["chef_client"]["winsw_dir"] do
    action :create
  end

  template "#{node["chef_client"]["winsw_dir"]}/chef-client.xml" do
    source "chef-client.xml.erb"
    notifies :run, "execute[restart chef-client using winsw wrapper]", :delayed
  end

  winsw_path = File.join(node["chef_client"]["winsw_dir"], node["chef_client"]["winsw_exe"])
  remote_file winsw_path do
    source node["chef_client"]["winsw_url"]
    not_if { File.exists?(winsw_path) }
  end

  # Work-around for CHEF-2541
  # Should be replaced by a service :restart action
  # in Chef 0.10.6
  execute "restart chef-client using winsw wrapper" do
    command "#{winsw_path} restart"
    not_if { WMI::Win32_Service.find(:first, :conditions => {:name => "chef-client"}).nil? }
    action :nothing
  end

  execute "Install chef-client service using winsw" do
    command "#{winsw_path} install"
    only_if { WMI::Win32_Service.find(:first, :conditions => {:name => "chef-client"}).nil? }
  end

  service "chef-client" do
    action :start
  end

when "launchd"

  version_checker = Chef::VersionConstraint.new(">= 0.10.10")
  mac_service_supported = version_checker.include?(node['chef_packages']['chef']['version'])

  if mac_service_supported
    template "/Library/LaunchDaemons/com.opscode.chef-client.plist" do
      source "com.opscode.chef-client.plist.erb"
      mode 0644
      variables(
        :launchd_mode => node["chef_client"]["launchd_mode"],
        :client_bin => client_bin
      )
    end

    service "chef-client" do
      service_name "com.opscode.chef-client"
      provider Chef::Provider::Service::Macosx
      action :start
    end
  else
    log("Mac OS X Service provider is only supported in Chef >= 0.10.10") { level :warn }
  end

when "bsd"
  log "You specified service style 'bsd'. You will need to set up your rc.local file."
  log "Hint: chef-client -i #{node["chef_client"]["client_interval"]} -s #{node["chef_client"]["client_splay"]}"
else
  log "Could not determine service init style, manual intervention required to start up the chef-client service."
end


attributes/default.rb
#
# Author:: Joshua Timberman (<joshua@opscode.com>)
# Author:: Seth Chisamore (<schisamo@opscode.com>)
# Cookbook Name:: chef
# Attributes:: default
#
# Copyright 2008-2011, Opscode, Inc
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'rbconfig'

default["chef_client"]["interval"]    = "1800"
default["chef_client"]["splay"]       = "20"
default["chef_client"]["log_dir"]     = "/var/log/chef"
default["chef_client"]["log_file"]    = nil
default["chef_client"]["log_level"]   = :info
default["chef_client"]["verbose_logging"] = true
default["chef_client"]["conf_dir"]    = "/etc/chef"
default["chef_client"]["bin"]         = "/usr/bin/chef-client"
default["chef_client"]["server_url"]  = ""
default["chef_client"]["validation_client_name"] = "chef-validator"
default["chef_client"]["cron"] = {
  "minute" => "0",
  "hour" => "*/4",
  "path" => nil,
  "environment_variables" => nil,
  "log_file" => "/dev/null",
  "use_cron_d" => false
}
default["chef_client"]["environment"] = nil
default["chef_client"]["load_gems"] = {}
default["chef_client"]["report_handlers"] = []
default["chef_client"]["exception_handlers"] = []
default["chef_client"]["checksum_cache_skip_expires"] = true
default["chef_client"]["daemon_options"] = []
default["ohai"]["disabled_plugins"] = [] #Sets disabled_plugins to empty array

case node['platform_family']
when "arch"
  default["chef_client"]["init_style"]  = "arch"
  default["chef_client"]["run_path"]    = "/var/run/chef"
  default["chef_client"]["cache_path"]  = "/var/cache/chef"
  default["chef_client"]["backup_path"] = "/var/lib/chef"
when "debian","rhel","fedora","suse"
  default["chef_client"]["init_style"]  = "init"
  default["chef_client"]["run_path"]    = "/var/run/chef"
  default["chef_client"]["cache_path"]  = "/var/cache/chef"
  default["chef_client"]["backup_path"] = "/var/lib/chef"
when "openbsd","freebsd"
  default["chef_client"]["init_style"]  = "bsd"
  default["chef_client"]["run_path"]    = "/var/run"
  default["chef_client"]["cache_path"]  = "/var/chef/cache"
  default["chef_client"]["backup_path"] = "/var/chef/backup"
# don't use bsd paths per COOK-1379
when "mac_os_x","mac_os_x_server"
  default["chef_client"]["init_style"]  = "launchd"
  default["chef_client"]["log_dir"]     = "/Library/Logs/Chef"
  # Launchd doesn't use pid files
  default["chef_client"]["run_path"]    = "/var/run/chef"
  default["chef_client"]["cache_path"]  = "/Library/Caches/Chef"
  default["chef_client"]["backup_path"] = "/Library/Caches/Chef/Backup"
  # Set to "daemon" if you want chef-client to run
  # continuously with the -d and -s options, or leave
  # as "interval" if you want chef-client to be run
  # periodically by launchd
  default["chef_client"]["launchd_mode"] = "interval"
when "openindiana","opensolaris","nexentacore","solaris2"
  default["chef_client"]["init_style"]  = "smf"
  default["chef_client"]["run_path"]    = "/var/run/chef"
  default["chef_client"]["cache_path"]  = "/var/chef/cache"
  default["chef_client"]["backup_path"] = "/var/chef/backup"
  default["chef_client"]["method_dir"] = "/lib/svc/method"
  default["chef_client"]["bin_dir"] = "/usr/bin"
when "smartos"
  default["chef_client"]["init_style"]  = "smf"
  default["chef_client"]["run_path"]    = "/var/run/chef"
  default["chef_client"]["cache_path"]  = "/var/chef/cache"
  default["chef_client"]["backup_path"] = "/var/chef/backup"
  default["chef_client"]["method_dir"] = "/opt/local/lib/svc/method"
  default["chef_client"]["bin_dir"] = "/opt/local/bin"
when "windows"
  default["chef_client"]["init_style"]  = "winsw"
  default["chef_client"]["conf_dir"]    = "C:/chef"
  default["chef_client"]["run_path"]    = "#{node["chef_client"]["conf_dir"]}/run"
  default["chef_client"]["cache_path"]  = "#{node["chef_client"]["conf_dir"]}/cache"
  default["chef_client"]["backup_path"] = "#{node["chef_client"]["conf_dir"]}/backup"
  default["chef_client"]["log_dir"]     = "#{node["chef_client"]["conf_dir"]}/log"
  default["chef_client"]["bin"]         = "C:/opscode/chef/bin/chef-client"
  #Required for minsw wrapper
  default["chef_client"]["ruby_bin"]    = File.join(RbConfig::CONFIG['bindir'], "ruby.exe")
  default["chef_client"]["winsw_url"]   = "http://repo1.maven.org/maven2/com/sun/winsw/winsw/1.9/winsw-1.9-bin.exe"
  default["chef_client"]["winsw_dir"]   = "C:/chef/bin"
  default["chef_client"]["winsw_exe"]   = "chef-client.exe"
else
  default["chef_client"]["init_style"]  = "none"
  default["chef_client"]["run_path"]    = "/var/run"
  default["chef_client"]["cache_path"]  = "/var/chef/cache"
  default["chef_client"]["backup_path"] = "/var/chef/backup"
end

default["chef_client"]["checksum_cache_path"] = "#{node["chef_client"]["cache_path"]}/checksums"
